File: JICS_MPI_11/Fortran/deadlockf/deadlockf.f
      program fixed
      implicit none
      include 'mpif.h'
      integer MSGLEN, ITAG_A, ITAG_B
      parameter (MSGLEN = 2048, ITAG_A = 100, ITAG_B = 200)
      real rmsg1(MSGLEN), rmsg2(MSGLEN)
      integer irank, idest, isrc, istag, iretag 
      integer ierr, I, request, irstatus(MPI_STATUS_SIZE)
      integer isstatus(MPI_STATUS_SIZE)

      call MPI_Init (ierr)
      call MPI_Comm_rank( MPI_COMM_WORLD, irank, ierr)
      
      do I = 1, MSGLEN
        rmsg1(I) = 100
        rmsg2(I) = -100
      end do
      
      if ( irank .eq. 0 ) then 
        idest  = 1
        isrc   = 1
        istag  = ITAG_A
        iretag = ITAG_B
      else if ( irank .eq. 1 ) then  
        idest  = 0
        isrc   = 0
        istag  = ITAG_B
        iretag = ITAG_A
      end if
      
      print *, "Task", irank, " has sent the message "
      call MPI_Isend( rmsg1, MSGLEN, MPI_REAL, idest, istag, 
     *  MPI_COMM_WORLD,  request, ierr)
      call MPI_Recv(rmsg2, MSGLEN, MPI_REAL, isrc, iretag, 
     *  MPI_COMM_WORLD, irstatus, ierr)
      call MPI_Wait(request, isstatus, ierr)
      print*, "Task", irank, " has received the message " 
      call MPI_Finalize (ierr)
      end

File: JICS_MPI_11/CPP/deadlockf/deadlockf.cpp
#include "mpi.h"
#include <iostream>

int main(int argc, char *argv[])		
{
#define MSGLEN 2048
  int ITAG_A = 100,ITAG_B = 200; 
  int irank, i, idest, isrc, istag, iretag;
  float rmsg1[MSGLEN];
  float rmsg2[MSGLEN];
  MPI::Status irstatus, isstatus;
  MPI::Request request;

  MPI::Init(argc, argv);
  irank = MPI::COMM_WORLD.Get_rank();

  for (i = 1; i <= MSGLEN; i++)
    {
      rmsg1[i] = 100;
      rmsg2[i] = -100;
    }
  if ( irank == 0 )
    { 
      idest  = 1;
      isrc   = 1;
      istag  = ITAG_A;
      iretag = ITAG_B;
    }
  else if ( irank == 1 )
    {
      idest  = 0;
      isrc   = 0;
      istag  = ITAG_B;
      iretag = ITAG_A;
    }

  cout << "Task "  << irank << " has sent the message" << endl;
  request = MPI::COMM_WORLD.Isend(rmsg1, MSGLEN, MPI::FLOAT, idest, istag); 
  MPI::COMM_WORLD.Recv(rmsg2, MSGLEN, MPI::FLOAT, isrc, iretag, irstatus);
  MPI_Wait(request, isstatus);
  cout << "Task " << irank << " has received the message" << endl;
  MPI::Finalize();
}
